src/knot_thing_main.c

```diff
@@ -6,196 +6,160 @@
 #include "knot_thing_main.h"
 
-static struct sensor_integer *isensor[KORE_INTEGER_SENSORS];
-static struct sensor_float *fsensor[KORE_FLOAT_SENSORS];
-static struct sensor_bool *bsensor[KORE_BOOL_SENSORS];
-static struct sensor_raw *rsensor[KORE_RAW_SENSORS];
-
-static int8_t isensor_num = 0;
-static int8_t fsensor_num = 0;
-static int8_t bsensor_num = 0;
-static int8_t rsensor_num = 0;
-
-/* Hash used to track value changes: stores last value hash */
-static int32_t ihash[KORE_INTEGER_SENSORS];
-static int32_t fhash[KORE_FLOAT_SENSORS];
-static int8_t bhash[KORE_BOOL_SENSORS];
-static int32_t rhash[KORE_RAW_SENSORS];
-
-static inline int32_t hash(int32_t value_int, int32_t multiplier,
-							int32_t value_dec)
+// TODO: normalize all returning error codes
+
+#define KNOT_THING_EMPTY_ITEM		"EMPTY ITEM"
+#define EMPTY_STRING			((uint8_t *)"")
+
+static struct {
+	// schema values
+	uint8_t			value_type;	// KNOT_VALUE_TYPE_* (int, float, bool, raw)
+	uint8_t			unit;		// KNOT_UNIT_*
+	uint16_t		type_id;	// KNOT_TYPE_ID_*
+	const char *		name;		// App defined data item name
+	// data values
+	knot_data_values	last_data;
+	uint8_t *		last_value_raw;
+	// config values
+	uint8_t			event_flags;	// Flags indicating when data will be sent
+	knot_data_values	lower_limit;
+	knot_data_values	upper_limit;
+	// Data read/write functions
+	knot_data_functions	functions;
+} data_items[KNOT_THING_DATA_MAX];
+
+static void reset_data_items(void)
 {
-	/* TODO: implement a better function */
-
-	return (value_int ^ multiplier ^ value_dec);
+	int8_t index = 0;
+
+	for (index = 0; index < KNOT_THING_DATA_MAX; index++)
+	{
+		data_items[index].name					= KNOT_THING_EMPTY_ITEM;
+		data_items[index].type_id				= KNOT_TYPE_ID_INVALID;
+		data_items[index].unit					= KNOT_UNIT_NOT_APPLICABLE;
+		data_items[index].value_type				= KNOT_VALUE_TYPE_INVALID;
+		data_items[index].event_flags				= KNOT_EVT_FLAG_UNREGISTERED;
+
+		data_items[index].last_data.value_f.multiplier		= 1; // as "last_data" is a union, we need just to
+		data_items[index].last_data.value_f.value_int		= 0; // set the "biggest" member
+		data_items[index].last_data.value_f.value_dec		= 0;
+		data_items[index].lower_limit.value_f.multiplier	= 1; // as "lower_limit" is a union, we need just to
+		data_items[index].lower_limit.value_f.value_int		= 0; // set the "biggest" member
+		data_items[index].lower_limit.value_f.value_dec		= 0;
+		data_items[index].upper_limit.value_f.multiplier	= 1; // as "upper_limit" is a union, we need just to
+		data_items[index].upper_limit.value_f.value_int		= 0; // set the "biggest" member
+		data_items[index].upper_limit.value_f.value_dec		= 0;
+		data_items[index].last_value_raw			= EMPTY_STRING;
+
+		data_items[index].functions.int_f.read	= NULL; // as "functions" is a union, we need just to
+		data_items[index].functions.int_f.write	= NULL; // set only one of its members
+	}
 }
 
-int8_t kore_init(void)
+int data_function_is_valid(knot_data_functions * func)
```

Comment:
nitpick, it should not block this code to be upstream: it should be *func (without space)